### PR TITLE
BUG: inconsistent behavior of DateOffset #47953

### LINKED ISSUE #47953: BUG: inconsistent behavior of DateOffset
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd

df = pd.DataFrame({
   'S':[pd.Timestamp('2019-04-30')],
   'A':[pd.DateOffset(months=1)]
   })

# FIRST TEST
>>> df['S'] + 26 * df['A']
0   2021-06-30
dtype: datetime64[ns]

# SECOND TEST
>>> df['S'].iloc[0] + 26 * df['A'].iloc[0]
Timestamp('2021-06-28 00:00:00')
```

### Issue Description

It seems like multiplying a DateOffset by constant leads to an inconsistent behavior: the first test gives 30/6 while the second 28/6 

### Expected Behavior

In any case it should be 2021-06-30

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 66e3805b8cabe977f40c05259cc3fcf7ead5687d
python           : 3.10.1.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19042
machine          : AMD64
processor        : Intel64 Family 6 Model 126 Stepping 5, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : Italian_Italy.1252

pandas           : 1.3.5
numpy            : 1.21.5
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 21.3.1
setuptools       : 58.1.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.8.0
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : 8.1.1
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.5.1
numexpr          : None
odfpy            : None
openpyxl         : 3.0.9
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : 1.7.3
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by rsm-23**
- [x] closes #47953
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v2.0.3.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by rsm-23** [comment]
@MarcoGorelli  can you please point the cause of build failure? I see tests failing, but they are running fine in my local. Also the other builds are passing.

**Comment by MarcoGorelli** [comment]
thanks for your PR - looks like some tests are failing, could you fix them up please?
```
FAILED pandas/tests/tseries/offsets/test_offsets.py::test_dateoffset_add_sub[offset_kwargs0-1970-01-01 00:00:00.000000001] - AssertionError
FAILED pandas/tests/tseries/offsets/test_offsets.py::test_dateoffset_add_sub[offset_kwargs1-1970-01-01 00:00:00.000000005] - AssertionError
FAILED pandas/tests/tseries/offsets/test_offsets.py::test_dateoffset_add_sub[offset_kwargs2-1969-12-31 23:59:59.999999999] - AssertionError
FAILED pandas/tests/tseries/offsets/test_offsets.py::test_dateoffset_add_sub_timestamp_with_nano - AssertionError
```

**Comment by rsm-23** [comment]
@MarcoGorelli the builds are passing now.

File: doc/source/whatsnew/v2.0.3.rst [hunk 1]
```diff
@@ -21,6 +21,7 @@ Fixed regressions
 
 Bug fixes
 ~~~~~~~~~
+- Bug in :class:`RelativeDeltaOffset` which caused inconsistent behavior upon multiplying a :class:`DateOffset` (:issue:`47953`)
```
Comment (MarcoGorelli) [inline doc/source/whatsnew/v2.0.3.rst:24]: `RelativeDeltaOffset` isn't technically part of the pandas API - could this be phrased just in terms of `DateOffset`?

File: pandas/tests/tseries/offsets/test_offsets.py [hunk 2]
```diff
@@ -1075,3 +1076,44 @@ def test_dateoffset_add_sub_timestamp_series_with_nano(offset, expected):
...
     testseries = offset + testseries
     assert testseries[0] == expected
+
+
+@pytest.mark.parametrize(
+    "month_value, scaling_factor, start_timestamp, expected_timestamp",
+    [(1, 2, "2020-01-30", "2020-03-30"), (2, 1, "2020-01-30", "2020-03-30")],
+)
+def test_offset_multiplication(
+    month_value, scaling_factor, start_timestamp, expected_timestamp
```
Comment (MarcoGorelli) [inline pandas/tests/tseries/offsets/test_offsets.py:1086]: how about `month_value` -> `n_months`?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Nice one @rsm-23 , thanks for working on this!

This is very close, just got two minor comments

File: pandas/tests/tseries/offsets/test_offsets.py (hunk 2 above)
Comment (rsm-23) [reply pandas/tests/tseries/offsets/test_offsets.py]: Changed the parameter name.

File: doc/source/whatsnew/v2.0.3.rst (hunk 1 above)
Comment (rsm-23) [reply doc/source/whatsnew/v2.0.3.rst]: Changed the phrasing. Please let me know if this works.

File: pandas/_libs/tslibs/offsets.pyx [hunk 3]
```diff
@@ -1221,12 +1221,7 @@ cdef class RelativeDeltaOffset(BaseOffset):
                 # perform calculation in UTC
                 other = other.replace(tzinfo=None)
 
-            if self.n > 0:
-                for i in range(self.n):
-                    other = other + self._offset
-            else:
-                for i in range(-self.n):
-                    other = other - self._offset
+            other = other + (self._offset * self.n)
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/offsets.pyx:1224]: sorry, I hadn't noticed this yesterday, but the code change here affects both `self.n > 0` and `self.n <= 0`. however, the test cases we've added only deal with the `self.n > 0` case

seeing as you've already parametrised the test (nice one!), would it be possible to add test cases for `n=0` and `n<0`? just to be safe

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v2.0.3.rst (hunk 1 above)
Comment (MarcoGorelli) [reply doc/source/whatsnew/v2.0.3.rst]: looks good!

File: pandas/_libs/tslibs/offsets.pyx (hunk 3 above)
Comment (rsm-23) [reply pandas/_libs/tslibs/offsets.pyx:1224]: My bad! I should have added all the scenarios. Thanks for pointing it out. Added them now! Do let me know if there's anything else. @MarcoGorelli

**Review by MarcoGorelli — APPROVED** [review]
Looks good to me on green, thanks @rsm-23 !

**Comment by rsm-23** [comment]
@MarcoGorelli can you please help me understand the next steps for this PR?

**Comment by MarcoGorelli** [comment]
I'll just leave this open a little to see if anyone else has comments/objections, and then will merge

**Comment by rsm-23** [comment]
Hi @MarcoGorelli thanks for approving the PR. Any updates on this? TiA!

File: doc/source/whatsnew/v2.0.3.rst [hunk 4]
```diff
@@ -21,6 +21,7 @@ Fixed regressions
 
 Bug fixes
 ~~~~~~~~~
+- Bug in :class:`DateOffset` which had inconsistent behavior when multiplying a :class:`DateOffset` object by a constant (:issue:`47953`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.0.3.rst:24]: Could you move this to `v2.1.0.rst`?

Comment (rsm-23) [reply doc/source/whatsnew/v2.0.3.rst]: Done @mroeschke

File: doc/source/whatsnew/v2.1.0.rst [hunk 5]
```diff
@@ -350,15 +342,14 @@ Bug fixes
...
 Categorical
 ^^^^^^^^^^^
-- Bug in :meth:`Series.astype` with ``dtype="category"`` for nullable arrays with read-only null value masks (:issue:`53658`)
 - Bug in :meth:`Series.map` , where the value of the ``na_action`` parameter was not used if the series held a :class:`Categorical` (:issue:`22527`).
 -
 
 Datetimelike
 ^^^^^^^^^^^^
 - :meth:`DatetimeIndex.map` with ``na_action="ignore"`` now works as expected. (:issue:`51644`)
+- Bug in :class:`DateOffset` which had inconsistent behavior when multiplying a :class:`DateOffset` object by a constant (:issue:`47953`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.1.0.rst:351]: This should be the only change in this file

Comment (rsm-23) [reply doc/source/whatsnew/v2.1.0.rst:362]: @mroeschke  My bad. Thanks for pointing out. I probably missed pulling the changes before pushing. Should be resolved now.

**Comment by MarcoGorelli** [comment]
right, let's ship this 🚀

**Pull request merged**